### PR TITLE
Make `cacheBefore` correctly handle allocation domain

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR fixes allocation domain handling of `cacheBefore`. In the future, I will validate `cacheAfter` and `cacheFork`. If they are not working, I will fix and add test for it. If they are already working, I will also add test for it.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/transform_replay.cpp [hunk 1]
```diff
@@ -519,29 +519,6 @@ std::pair<TensorDomain*, size_t> TransformReplay::replayPasC(
     }
   }
 
-  if (producer->hasAllocation()) {
```
Comment (zasdfgbnm) [inline csrc/transform_replay.cpp:522]: This check is removed. If the replay makes the allocation domain no longer valid, then the ctor of `TensorDomain` will detect that.

File: csrc/tensor_view.cpp [hunk 2]
```diff
@@ -1229,8 +1232,33 @@ TensorView* TensorView::cacheBefore(LoadStoreOpType cache_op) {
 
   auto replayed_consumer_pair =
       TransformReplay::replayCasP(consumer, producer, -1);
+
   consumer->setDomain(replayed_consumer_pair.first);
 
+  // Recover allocation domain from transform replay
+  // TODO: should move this to
```
Comment (naoyam) [inline csrc/tensor_view.cpp:1239]: The following code looks good to me, but I'm not sure what you're saying with this comment.

**Review by naoyam — APPROVED** [review]
LGTM

File: csrc/tensor_view.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/tensor_view.cpp]: Updated the comment. Basically the logic below is just equivalent to asking `replayCasP` to replay the allocation domain of the consumer consistent with the producer. So we should consider making it an opt-in functionality of `replayCasP` if there are other use cases needing this.

Comment (naoyam) [reply csrc/tensor_view.cpp]: So, `replayCasP` also set its allocation domain as the producer transformations are replayed on the consumer? Having that option seems to make sense.

**Pull request merged**